annotation/tests/test_tasks_services.py

```diff
@@ -585,3 +636,1697 @@ def test_filter_tasks_db_no_files_or_jobs(
+    with patch(
+        "annotation.tasks.services.AgreementScoreComparingResult"
+    ) as mock:
+        yield mock
+
+
+def test_compare_agreement_scores_all_above_min_match(
+    mock_agreement_score_response: Mock,
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.8
+    mock_parse_obj_as.return_value = [MagicMock(spec=ResponseScore)]
+    mock_task_metric.side_effect = (
+        lambda task_from_id, task_to_id, metric_score: MagicMock(
+            spec=TaskMetric,
+            task_from_id=task_from_id,
+            task_to_id=task_to_id,
+            metric_score=metric_score,
+        )
+    )
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=True,
+        task_metrics=[
+            mock_task_metric(1, 2, 0.95),
+            mock_task_metric(1, 3, 0.8),
+            mock_task_metric(2, 3, 0.85),
+        ],
+    )
+    result = compare_agreement_scores(mock_agreement_score_response, min_match)
+    assert result.agreement_score_reached
+
+
+def test_compare_agreement_scores_some_below_min_match(
+    mock_agreement_score_response: Mock,
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.9
+    mock_parse_obj_as.return_value = [MagicMock(spec=ResponseScore)]
+    mock_task_metric.side_effect = (
+        lambda task_from_id, task_to_id, metric_score: MagicMock(
+            spec=TaskMetric,
+            task_from_id=task_from_id,
+            task_to_id=task_to_id,
+            metric_score=metric_score,
+        )
+    )
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=False,
+        task_metrics=[mock_task_metric(1, 2, 0.95)],
+    )
+    result = compare_agreement_scores(mock_agreement_score_response, min_match)
+    assert not result.agreement_score_reached
+
+
+def test_compare_agreement_scores_empty_response(
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.5
+    mock_parse_obj_as.return_value = []
+    mock_task_metric.return_value = MagicMock(spec=TaskMetric)
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=False,
+        task_metrics=[],
+    )
+    result = compare_agreement_scores([], min_match)
+    assert not result.agreement_score_reached
+    assert result.task_metrics == []
+
+
+def test_save_agreement_metrics(mock_session: Mock):
+    mock_scores = MagicMock()
+    mock_scores.task_metrics = []
+    save_agreement_metrics(mock_session, mock_scores)
```

Comment:
good example of testing mocks - you dont test the real functionality. Simply create the real AgreementScoreComparingResult object and check bulk_save_objects call **real** arguments. It actually applies to a lot of your tests